Infrastructure/src/Linux32.cpp

```diff
@@ -0,0 +1,208 @@
+    x.size = size;
+    return x;
+}
+
+void Linux32::virtualFree(void* ptr, size_t size){
+    munmap(ptr, size);
+    return;
+}
+
+WindowPtr Linux32::createWindow(const u32 width, const u32 height, const std::string &title){
+    return nullptr;
+}
+
+real64 Linux32::getTime(){
+    timespec ts;
+    clock_gettime(CLOCK_REALTIME, &ts);
+    return (double)(((long) ts.tv_sec * NANOSECONDS_PER_SECOND) + ts.tv_nsec)/((double)(NANOSECONDS_PER_SECOND));
+}
+
+int Linux32::patestCallback(const void* inputBuffer, void* outputBuffer,
+                   unsigned long framesPerBuffer,
+                   const PaStreamCallbackTimeInfo* timeInfo,
+                   PaStreamCallbackFlags statusFlags,
+                   void* userData)
+{
+    GetAudioCallback m_pullAudio = reinterpret_cast<GetAudioCallback>(userData);
+
+    if(m_pullAudio)
+        m_pullAudio(static_cast<byte*>(outputBuffer),framesPerBuffer,userData);
+
+    return paContinue;
+}
+
+void Linux32::initSound(AudioFormat format){
+    PaStreamParameters output_parameters;
+    PaError err;
+
+    err = Pa_Initialize();
+    if(err != paNoError)
+        throw ServiceException("Failed to init sound");
+
+    auto numDevices = Pa_GetDeviceCount();
+    printf("numDevices = %d\n",numDevices);
+    const   PaDeviceInfo *deviceInfo;
+    int i;
+    for(i=0;i<numDevices;i++){
+        deviceInfo = Pa_GetDeviceInfo(i);
+        printf("name = %s\n",deviceInfo->name);
+    }
+    output_parameters.device = Pa_GetDefaultOutputDevice();
+    //output_parameters.device = 4;
+    printf("i = %d\n",i);
+    if(output_parameters.device == paNoDevice)
+        throw ServiceException("OS can't find audio device");
+
+    output_parameters.channelCount = format.Channels;
+    output_parameters.sampleFormat = paInt16;
+    output_parameters.suggestedLatency = Pa_GetDeviceInfo(output_parameters.device)->defaultLowOutputLatency;
+    output_parameters.hostApiSpecificStreamInfo = NULL;
+
+    err = Pa_OpenStream(
+        &stream,
+        NULL,
+        &output_parameters,
+        format.SamplesPerSec,
+        512,
+        paClipOff,
+        patestCallback,
+        (void*)m_pullAudio
+        );
+
+    if(err != paNoError)
+        throw ServiceException("OS can't open audio output stream");
+
+    err = Pa_StartStream(stream);
+
+    if(err != paNoError)
+        throw ServiceException("OS can't start audio stream");
+    return;
+}
+
+void Linux32::setPullAudioCallback(GetAudioCallback function){
+    m_pullAudio = function;
+    return;
+}
+
+bool Linux32::fileExists(const std::string& fileName){
+    if (access( fileName.c_str(), F_OK ) != -1)
+        return true;
+    return false;
+}
+
+void* Linux32::openFile(const std::string& fileName, const AccessPermission permission, u64& out_size){
+    return nullptr;
+}
+
+void* Linux32::createFile(const std::string& fileName, const AccessPermission permission){
+    std::ofstream* file = new std::ofstream(fileName);
```

Comment:
I think I can still do all these functions no problem. So let's have a final word about it :D I say it doesn't matter I can still do all you need, but you're the project boss, if you want me to use the OS functions I'll use them.

So decide and let me know.